Sources/Clients/OnLedgerEntitiesClient/OnLedgerEntitiesClient+Live.swift

```diff
@@ -6,101 +6,393 @@ import SharedModels
+			nonFungibleIds: ids
+		))
+
+		return .init(tokens: tokens, nextPageCursor: nextPageCursor)
+	}
+
+	static func fetchNonFungibleData(_ request: GetNonFungibleTokenDataRequest) async throws -> [OnLedgerEntity] {
 		@Dependency(\.gatewayAPIClient) var gatewayAPIClient
+		let response = try await request.nonFungibleIds
+			.chunks(ofCount: maximumNFTIDChunkSize)
+			.parallelMap { ids in
+				try await gatewayAPIClient.getNonFungibleData(.init(
+					atLedgerState: request.atLedgerState?.selector,
+					resourceAddress: request.resource.address,
+					nonFungibleIds: Array(ids.map { try $0.localId().toString() })
+				))
+			}
 
-		return try await gatewayAPIClient
-			.getEntityDetails(addresses.map(\.address), .resourceMetadataKeys, nil)
-			.items
-			.compactMap(createEntity)
+		return try response
+			.flatMap { item in
+				try item.nonFungibleIds.map { id in
+					try OnLedgerEntity.nonFungibleToken(.init(
+						id: .fromParts(
+							resourceAddress: .init(address: request.resource.address),
+							nonFungibleLocalId: .from(stringFormat: id.nonFungibleId)
+						),
+						data: id.details
+					))
+				}
+			}
 	}
 
 	@Sendable
-	static func createEntity(from item: GatewayAPI.StateEntityDetailsResponseItem) throws -> OnLedgerEntity? {
-		let dappDefinitions = item.explicitMetadata?.dappDefinitions?.compactMap { try? DappDefinitionAddress(validatingAddress: $0) }
-
-		switch item.details {
-		case let .fungibleResource(fungibleDetails):
-			return try .resource(.init(
-				resourceAddress: .init(validatingAddress: item.address),
-				divisibility: fungibleDetails.divisibility,
-				name: item.explicitMetadata?.name,
-				symbol: item.explicitMetadata?.symbol,
-				description: item.explicitMetadata?.description,
-				iconURL: item.explicitMetadata?.iconURL,
-				behaviors: item.details?.fungible?.roleAssignments.extractBehaviors() ?? [],
-				tags: item.explicitMetadata?.extractTags() ?? [],
-				totalSupply: try? RETDecimal(value: fungibleDetails.totalSupply),
-				dappDefinitions: dappDefinitions
-			))
-		case let .nonFungibleResource(nonFungibleDetails):
-			return try .resource(.init(
-				resourceAddress: .init(validatingAddress: item.address),
-				divisibility: nil,
-				name: item.explicitMetadata?.name,
-				symbol: nil,
-				description: item.explicitMetadata?.description,
-				iconURL: item.explicitMetadata?.iconURL,
-				behaviors: item.details?.nonFungible?.roleAssignments.extractBehaviors() ?? [],
-				tags: item.explicitMetadata?.extractTags() ?? [],
-				totalSupply: try? RETDecimal(value: nonFungibleDetails.totalSupply),
-				dappDefinitions: dappDefinitions
-			))
+	static func fetchEntitiesWithCaching(
+		for identifiers: [CacheClient.Entry.OnLedgerEntity],
+		forceRefresh: Bool = false,
+		refresh: (_ identifiers: [CacheClient.Entry.OnLedgerEntity]) async throws -> [OnLedgerEntity]
+	) async throws -> [OnLedgerEntity] {
+		@Dependency(\.cacheClient) var cacheClient
+
+		if forceRefresh {
```

Comment:
maybe `guard !forceRefrech else ` would make the non forceRefesh scope easier to read.